server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/CellIdSource.java

```diff
@@ -81,20 +87,32 @@ public SortedBinaryDocValues bytesValues(LeafReaderContext ctx) {
         private MultiGeoPointValues geoValues;
         private int precision;
         private GeoPointLongEncoder encoder;
+        private org.elasticsearch.common.geo.GeoPoint topLeft;
+        private org.elasticsearch.common.geo.GeoPoint bottomRight;
 
-        protected CellValues(MultiGeoPointValues geoValues, int precision, GeoPointLongEncoder encoder) {
+        protected CellValues(MultiGeoPointValues geoValues, int precision, org.elasticsearch.common.geo.GeoPoint topLeft,
+                             org.elasticsearch.common.geo.GeoPoint bottomRight, GeoPointLongEncoder encoder) {
             this.geoValues = geoValues;
             this.precision = precision;
             this.encoder = encoder;
+            this.topLeft = topLeft;
+            this.bottomRight = bottomRight;
+            assert Double.isNaN(topLeft.lon()) ==
+                ((Double.isNaN(topLeft.lat()) ^ Double.isNaN(bottomRight.lon()) ^ Double.isNaN(bottomRight.lat())));
         }
 
         @Override
         public boolean advanceExact(int docId) throws IOException {
             if (geoValues.advanceExact(docId)) {
                 resize(geoValues.docValueCount());
+                int j = 0;
                 for (int i = 0; i < docValueCount(); ++i) {
                     org.elasticsearch.common.geo.GeoPoint target = geoValues.nextValue();
-                    values[i] = encoder.encode(target.getLon(), target.getLat(), precision);
```

Comment:
another question I have @iverase is how it might be best to incorporate these bounds during the rasterization of shapes (for the geoshape-doc-values branch). It seems like there can be a few different situations that should be concerned differently. I am thinking it makes sense to pass the bounds to the GeoGridTiler and use it to clip the tile the tile-queries such that tiles that intersect the bounds are made smaller to their overlapping shape. tile-queries that are disjoint from the bounds are immediately ignored and no `relate` call is made. If this works, I think it would be nice since then there are no additional calls to relate in any pre-processing way and it does not complicate the tiling logic too much. What do you think?

Reply:
We will have to check performance but my first the on this would be first to check if the shape intersects the provided bounding box. If true then go to the tiler and again filter out the tiles that do not intersects the provided bounds.

Reply:
Actually we can skip the first?

Reply:
I will check, but even if we do the initial intersection check, we need to clip each tile according to the provided bounds. That is why I thought the initial check would not be too helpful most times that the bounds provided to the geotile_grid match that which is passed to the geo_bounding_box filter.